Meadowlark-js/packages/meadowlark-core/src/handler/DocumentReferenceExtractor.ts

```diff
@@ -109,93 +106,81 @@ function extractDocumentBodyPaths(referenceComponents: ReferenceComponent[], ent
  * for all of the portions of the reference.
  *
- * Example of a foreign key bundle, representing a collection of two references to the same entity type
+ * Example of document identities, representing a collection of two references to the same entity type
  * [
  *   ['classPeriodName=z1', 'schoolId=24', 'studentId=333'],
  *   ['classPeriodName=z2', 'schoolId=25', 'studentId=444']
  * ]
  */
-function foreignKeyBundlesFromReferenceGroup(
+function documentIdentitiesFromReferenceGroup(
   referenceGroup: ReferenceGroup,
-  body: object,
+  document: object,
   entity: TopLevelEntity,
-): ExtractedValue[][] {
-  // The body paths of the reference group, which matches the API document body paths to the
+): DocumentIdentity[] {
+  // The document paths of the reference group, which matches the API document body paths to the
   // foreign key elements of the entity being referenced.
-  const documentBodyPaths: string[] = extractDocumentBodyPaths(referenceGroup.referenceComponents, entity);
+  const documentPaths: string[] = documentPathsFromReferenceComponents(referenceGroup.referenceComponents, entity);
 
-  const bodyPathSets: string[][][] = extractBodyPaths(referenceGroup, body);
+  const documentPathSets: string[][][] = extractRawDocumentPaths(referenceGroup, document);
   // if lengths are different, something optional was not present, so we are done
-  if (documentBodyPaths.length !== bodyPathSets.length) return [];
+  if (documentPaths.length !== documentPathSets.length) return [];
 
-  const orderedAndGroupedByEnding: ExtractedValue[][] = R.zipWith(
-    (namePath: string, bodyPathSet: string[][]) =>
-      bodyPathSet.map((bodyPath) => ({ namePath, bodyValue: R.path(bodyPath, body) as string })),
-    documentBodyPaths,
-    bodyPathSets,
+  const orderedAndGroupedByEnding: DocumentElement[][] = R.zipWith(
+    (name: string, documentPathSet: string[][]) =>
+      documentPathSet.map((documentPath) => ({ name, value: R.path(documentPath, document) as string })),
+    documentPaths,
+    documentPathSets,
   );
   // example result for orderedAndGroupedByEnding after R.zipWith():
   // [
-  //   [{namePath: 'classPeriodName', value: 'z1'}, {namePath: 'classPeriodName', value: 'z2'}],
-  //   [{namePath: 'schoolId', value: '24'}, {namePath: 'schoolId', value: '25'}],
-  //   [{namePath: 'studentId', value: '333'}, {namePath: 'studentId', value: '444'}]
+  //   [{name: 'classPeriodName', value: 'z1'}, {name: 'classPeriodName', value: 'z2'}],
+  //   [{name: 'schoolId', value: '24'}, {name: 'schoolId', value: '25'}],
+  //   [{name: 'studentId', value: '333'}, {name: 'studentId', value: '444'}]
   // ]
 
   if (orderedAndGroupedByEnding.length === 0) return [];
 
-  const foreignKeyBundles: ExtractedValue[][] = multiZip(...orderedAndGroupedByEnding);
-  // example result for foreignKeyBundles after multiZip():
+  const documentIdentities: DocumentIdentity[] = multiZip(...orderedAndGroupedByEnding);
```

Comment:
An example of where document identities are a way better concept that 2-dimensional arrays called "foreign key bundles" 😆